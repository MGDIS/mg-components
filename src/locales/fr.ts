const locale = {
  "errors": {
    "required": "Ce champs est obligatoire.",
    "date": {
      "badInput": "Le format de la date doit être du type <span aria-hidden=\"true\">jj/mm/aaaa</span><span class=\"sr-only\">j j / m m / a a a a</span> et supérieure au 01/01/1900"
    }
  },
  "nbCharLeft" : "{counter} caractères disponibles.",
  "input": {
    "select" : {
      "placeholder": "Veuillez sélectionner une valeur"
    },
    "checkbox": {
      "true": "Oui",
      "false": "Non",
      "indeterminate": "Indéterminé",
    }
  }
}

export default locale;
