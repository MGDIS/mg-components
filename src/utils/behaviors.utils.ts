import { isMgActionMore } from '../components/molecules/menu/mg-action-more/mg-action-more.conf';
import { isMgMenuItem, Status } from '../components/molecules/menu/mg-menu-item/mg-menu-item.conf';

export enum OverflowBehaviorElements {
  MORE = 'data-overflow-more',
  BASE_INDEX = 'data-overflow-base-index',
  PROXY_INDEX = 'data-overflow-proxy-index',
}

export class OverflowBehavior {
  // variables
  private resizeObserver: ResizeObserver;
  private moreELement: HTMLElement | HTMLMgActionMoreElement;
  private children: Element[];

  constructor(private element: Element, private render: () => HTMLMgMenuItemElement | HTMLMgActionMoreElement) {}

  /**********
   * Public *
   **********/

  /**
   * init behavior
   *
   * @returns {void}
   */
  init = (): void => {
    if (!this.resizeObserver) {
      // add resize observer
      this.resizeObserver = new ResizeObserver(entries => {
        entries.forEach(entry => {
          this.run(entry.contentRect.width);
        });
      });
      this.resizeObserver.observe(this.element);
    }
  };

  /**
   * Disconnect ResizeObserver
   *
   * @returns {void}
   */
  disconnect = (): void => {
    this.resizeObserver?.disconnect();
  };

  /************
   * Internal *
   ************/

  /**
   * run overflow behavior
   *
   * @param {number} width width of the container
   * @returns {void}
   */
  private run = (width: number): void => {
    if (!this.moreELement) {
      // set moreElement
      this.moreELement = this.render();
      this.moreELement.setAttribute(OverflowBehaviorElements.MORE, '');

      // set children
      this.children = Array.from(this.element.children);
      let moreELementChildren;
      const filter = ['MG-MENU-ITEM', 'MG-BUTTON'];
      if (isMgMenuItem(this.moreELement)) {
        moreELementChildren = Array.from(this.moreELement.querySelector('mg-menu').children).filter(element => filter.includes(element.nodeName));
      } else {
        moreELementChildren = Array.from(this.moreELement.children).filter(element => filter.includes(element.nodeName));
      }

      this.children.forEach((child, index) => {
        child.setAttribute(OverflowBehaviorElements.BASE_INDEX, `${index}`);
        if (child.getAttribute(OverflowBehaviorElements.MORE) === null) {
          // set moreElement children
          moreELementChildren[index].setAttribute(OverflowBehaviorElements.PROXY_INDEX, `${index}`);
        }
      });
    }

    this.updateDisplayedItems(width);
    if (isMgMenuItem(this.moreELement)) this.updateActiveStatus();
  };

  /**
   * Update displayed items in container from a given available space
   *
   * @param {number} availableWidth available width in container to display child items
   * @returns {void}
   */
  private updateDisplayedItems = (availableWidth: number): void => {
    this.restoreItems();
    this.children.reduce(
      (acc, curr: HTMLMgMenuItemElement) => {
        acc.accWidth += curr.offsetWidth;
        const isPreviousItemHidden = acc.previousItem !== null && acc.previousItem.getAttribute('hidden') !== null;
        // if previous item is hidden AND is NOT more element we hidde current
        // OR if item has an overflow we hidde current
        // OR if current item is more element AND have NOT previous hidden items, current more element item is an hidden item
        // ELSE current is a display item
        if (
          (isPreviousItemHidden && !this.isMoreElement(curr)) ||
          this.isOverflowElement(acc.accWidth, curr, availableWidth) ||
          (this.isMoreElement(curr) && !isPreviousItemHidden)
        ) {
          this.toggleItem(curr, true);
        } else this.toggleItem(curr, false);

        // update previous item with current item
        acc.previousItem = curr;

        return acc;
      },
      { accWidth: 0, previousItem: null },
    );
  };

  /**
   * Utility method to know if item overflow partialy or totaly in the container
   *
   * @param {number} cumulateWidth previous sibling elements cumulate width + item width
   * @param {HTMLElement} item item HTML element
   * @param {number} availableWidth container available width
   * @returns {boolean}
   */
  private isOverflowElement = (cumulateWidth: number, item: HTMLElement, availableWidth: number): boolean => {
    if (item.previousElementSibling === null) return false;
    else if (this.isMoreElement(item)) return false;
    else if (item.previousElementSibling !== null && item.nextElementSibling !== null && !this.isMoreElement(item.nextElementSibling))
      return cumulateWidth + this.moreELement.offsetWidth > availableWidth;
    else return cumulateWidth > availableWidth;
  };

  /**
   * Utility method to know if a given element is the 'MORE' element
   *
   * @param {Element} element element to match with 'MORE' element
   * @returns {boolean}
   */
  private isMoreElement = (element: Element): boolean => element.getAttribute(OverflowBehaviorElements.MORE) !== null;

  /**
   * Toggle display item element
   *
   * @param {HTMLElement} item item to control
   * @param {boolean} isHidden is item hidden
   * @returns {void}
   */
  private toggleItem = (item: HTMLElement, isHidden: boolean): void => {
    const mgActionMenuindex = Number(item.getAttribute(OverflowBehaviorElements.BASE_INDEX));
    if (this.isMoreElement(item)) this.toggleElement(this.moreELement, isHidden);
    else if (mgActionMenuindex >= 0) {
      this.toggleElement(item, isHidden);
      if (isMgActionMore(this.moreELement)) {
        this.moreELement.items[mgActionMenuindex].hidden = !isHidden;
        // reset moreElement items to handle component watcher
        this.moreELement.items = [...this.moreELement.items];
      } else {
        this.toggleElement(this.moreELement.querySelector(`[${OverflowBehaviorElements.PROXY_INDEX}="${mgActionMenuindex}"]`), !isHidden);
      }
    }
  };

  /**
   * Toggle element's hidde attribute
   *
   * @param {Element} element element to toggle
   * @param {boolean} isHidden element is hidden
   * @returns {void}
   */
  private toggleElement = (element: Element, isHidden: boolean): void => {
    if (isHidden) element.setAttribute('hidden', '');
    else element.removeAttribute('hidden');
  };

  /**
   * Remove hidden attributes on all items
   *
   * @returns {void}
   */
  private restoreItems = (): void => {
    this.children.forEach(item => item.removeAttribute('hidden'));
  };

  /**
   * Update more element status if active child
   *
   * @returns {void}
   */
  private updateActiveStatus = (): void => {
    const hasActiveChild = Array.from(this.moreELement.querySelector('mg-menu').children).find(
      child => child.getAttribute('status') === Status.ACTIVE && child.getAttribute('hidden') === null,
    );
    this.moreELement.setAttribute('status', hasActiveChild ? Status.ACTIVE : Status.VISIBLE);
  };
}
