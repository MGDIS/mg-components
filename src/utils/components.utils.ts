/**
 * Create random ID
 * @param prefix {string} add prefix to created ID
 * @returns {string} ID
 */
export function createID(prefix?: string): string {
  return (typeof prefix === 'string' && prefix !== '' ? `${prefix}-` : '' ) + Math.random().toString(36).substr(2, 10);
}

/**
 * Class to manage component classlist
 * Set() are not working when imported in project
 */
export class ClassList {
  classes: string[];

  constructor (classlist:string[] = []) {
    this.classes = classlist;
  }

  /**
   * Add class
   * @param className
   * @returns {void}
   */
  add = (className:string) => {
    if(!this.has(className)) {
      this.classes.push(className);
    }
  }

  /**
   * Delete class
   * @param className
   * @returns {void}
   */
  delete = (className:string) => {
    const index = this.classes.indexOf(className);
    if(index > -1) {
      this.classes.splice(index, 1);
    }
  }

  /**
   * Check if class exist in list
   * @param className
   * @returns {boolean}
   */
  has = (className:string) => {
    return this.classes.includes(className);
  }

  /**
   * Join classes seperated by spaces
   * @returns {string}
   */
  join = ():string => {
    return this.classes.join(' ');
  }
}
