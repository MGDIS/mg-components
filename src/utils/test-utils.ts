import { newE2EPage, E2EPage } from "@stencil/core/testing"
import { Page as PuppeteerPage } from "puppeteer"

export type DesignSystemE2EPage = E2EPage & Pick<PuppeteerPage, "screenshot" | "viewport">

type DesignSystemE2EPageOptions = { html: string; viewportWidth: number }

export async function createPage(optionsOrHtml?: string | DesignSystemE2EPageOptions) {
  const options: DesignSystemE2EPageOptions =
    typeof optionsOrHtml === "string" ? { html: `<link rel="stylesheet" href="http://localhost:3333/build/design-system.css" />${optionsOrHtml}`, viewportWidth: 600 } : optionsOrHtml

  const page = (await newE2EPage()) as DesignSystemE2EPage
  const viewport = Object.assign({ height: page.viewport().height }, { width: options.viewportWidth })
  await page.setViewport(viewport)
  await page.setContent(options.html, { waitUntil: "networkidle0" })
  // await page.evaluateHandle(() => (document as any).fonts.ready)

  // monkey patch screenshot function to add some extra features
  const screenshot = page.screenshot
  page.screenshot = async function() {
    // get the element's height, and set viewport to that height
    // this enables us to get full page, clipped screenshots
    const htmlElement = await page.$("html")
    const { width, height } = await htmlElement.boundingBox()
    await page.setViewport({ width: page.viewport().width, height: Math.round(height) })

    return screenshot.call(page, {
      clip: {
        x: 0,
        y: 0,
        width: Math.round(width),
        height: Math.round(height),
      },
    })
  }

  return page
}
