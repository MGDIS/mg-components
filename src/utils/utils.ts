/**
 * Create random ID
 * @param prefix {string} add prefix to created ID
 * @returns {string} ID
 */
export function createID(prefix?: string): string {
  return (typeof prefix === 'string' && prefix !== '' ? `${prefix}-` : '' ) + Math.random().toString(36).substr(2, 10);
}

/**
 * Change date format
 * @param date {string} date with format yyyy-mm-dd
 * @returns {string} date with format dd/mm/yyyy
 */
export function formatDate(date: string): string {
  const dateRegexp = /\d{4}-\d{2}-\d{2}/;
  if(typeof date !== 'string' || date === '' || !dateRegexp.test(date)) {
    return '';
  }
  const d:Date = new Date(date);
  return `${d.getDate()}/${d.getMonth()+1}/${d.getFullYear()}`;
}

/**
 * Class to manage component classlist
 * Set() are not working when imported in project
 */
export class ClassList {
  classes: string[];

  constructor (classlist:string[] = []) {
    this.classes = classlist;
  }

  /**
   * Add class
   * @param className
   * @returns {void}
   */
  add = (className:string) => {
    if(!this.classes.includes(className)) {
      this.classes.push(className);
    }
  }

  /**
   * Delete class
   * @param className
   * @returns {void}
   */
  delete = (className:string) => {
    const index = this.classes.indexOf(className);
    if(index > -1) {
      this.classes.splice(index, 1);
    }
  }

  /**
   * Join classes seperated by spaces
   * @returns {string}
   */
  join = ():string => {
    return this.classes.join(' ');
  }
}
