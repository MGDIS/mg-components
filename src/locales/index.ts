import en from './en/messages.json';
import fr from './fr/messages.json';

const defaultLocale = 'en';
const messages = { en, fr };

/**
 * Get locale
 *
 * @param {HTMLElement} element element that needs to know the locale to use
 * @returns {string} locale
 */
const getLocale = (element: HTMLElement): string => {
  return (element.closest('[lang]') as HTMLElement)?.lang || navigator.language || defaultLocale;
};

/**
 * Get Intl object
 * We load the defined locale but for now we only support the first subtag for messages
 *
 * @param {HTMLElement} element element we need to get the language
 * @returns {{ locale: string; messages: unknown }} messages object
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const initLocales = (element: HTMLElement): { locale: string; messages: any } => {
  // Get local
  const locale = getLocale(element);
  // Only keep first subtag
  const localeSubtag = locale.split('-').shift();
  // Return
  return {
    locale,
    messages: messages[localeSubtag] || messages[defaultLocale],
  };
};
