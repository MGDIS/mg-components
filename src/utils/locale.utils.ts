import { locale, currency } from "../locales";

/**
 * Format number to the locale currency
 * @param number {number}
 * @returns {string} formatted currency
 */
export function localeCurrency(number:number):string {
  return new Intl.NumberFormat(locale, {style: "currency", currency}).format(number);
};

/**
 * Locale date format
 * @param date {string}
 * @returns {string} formatted date
 */
export function localeDate(date:string):string {
  const dateRegexp = /\d{4}-\d{2}-\d{2}/;
  if(typeof date !== 'string' || date === '' || !dateRegexp.test(date)) {
    return '';
  }
  return new Intl.DateTimeFormat(locale).format(new Date(date));
}
