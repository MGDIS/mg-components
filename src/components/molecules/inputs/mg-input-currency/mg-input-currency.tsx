import { Component, Element, Event, h, Prop, EventEmitter, State, Watch } from '@stencil/core';
import { MgInput } from '../MgInput';
import { createID, ClassList } from '../../../../utils/components.utils';
import { messages } from '../../../../locales';
import { localeCurrency } from '../../../../utils/locale.utils';

@Component({
  tag: 'mg-input-currency',
  styleUrl: 'mg-input-currency.scss',
  shadow: true,
})
export class MgInputCurrency {
  /************
   * Internal *
   ************/

  // Manage different values
  private storedValue: string;
  private numericValue: number;
  private readonlyValue: string;
  // Classes
  private classFocus: string = 'is-focused';
  private classError: string = 'is-not-valid';

  /**************
   * Decorators *
   **************/

  @Element() element: HTMLMgInputCurrencyElement;

  /**
   * Component value
   */
  @Prop({ mutable: true, reflect: true }) value: string;
  @Watch('value')
  validateValue(newValue: string) {
    if (newValue !== undefined && newValue !== null) {
      // Filter input
      const [number, decimal = ''] = newValue.replace('-', '').split(/[\.,]/);
      if (newValue === '' || (newValue.match(/^[\-]?\d+[.,]?\d*$/) && number.length <= 13 && decimal.length <= 2)) {
        this.storedValue = newValue;
      } else if (this.storedValue !== undefined) {
        newValue = this.storedValue;
      }
      else {
        newValue = null;
      }
      // Set value and input value
      this.value = newValue;
      const elementInput = this.element.shadowRoot.getElementById(this.identifier) as HTMLInputElement;
      if (elementInput !== null) elementInput.value = this.value;
      // emit numeric value
      this.numericValue = this.value !== '' && this.value !== null ? parseFloat(this.value.replace(',', '.')) : null;
      this.valueChange.emit(this.numericValue);
      // Set readOnlyValue
      this.readonlyValue = localeCurrency(this.numericValue);
    }
  }

  /**
   * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
   * If not set, it will be created.
   */
  @Prop() identifier?: string = createID('mg-input-currency');

  /**
   * Input name
   * If not set the value equals the identifier
   */
  @Prop() name?: string = this.identifier;

  /**
   * Input label
   * Required
   */
  @Prop() label!: string;

  /**
   * Define if label is displayed on top
   */
  @Prop() labelOnTop: boolean;

  /**
   * Define if label has colon ":"
   */
  @Prop() labelColon: boolean = false;

  /**
   * Input placeholder.
   * It should be a word or short phrase that demonstrates the expected type of data, not a replacement for labels or help text.
   *
   */
  @Prop() placeholder: string;

  /**
   * Define if input is required
   */
  @Prop() required: boolean = false;

  /**
   * Define if input is readonly
   */
  @Prop() readonly: boolean = false;

  /**
   * Define if input is disabled
   */
  @Prop() disabled: boolean = false;

  /**
   * Add a tooltip message next to the input
   */
  @Prop() tooltip: string;

  /**
   * Template to use for characters left sentence
   */
  @Prop() helpText: string;

  /**
   * Maximum value
   */
  @Prop() max: number;

  /**
   * Minimum value
   */
  @Prop() min: number;

  /**
   * Define input pattern to validate
   */
  @Prop({ mutable: true, reflect: true }) valid: boolean;

  /**
   * Define input pattern error message
   */
  @Prop({ mutable: true, reflect: true }) invalid: boolean;

  /**
   * Component classes
   */
  @State() classList: ClassList = new ClassList(['mg-input--currency']);

  /**
   * Error message to display
   */
  @State() errorMessage: string;

  /**
   * Displayed value in input
   * Change on focus/blur
   */
  @State() displayValue: string;

  /**
   * Emmited event when value change
   */
  @Event() valueChange: EventEmitter<number>;

  /**
   * Handle input event
   * @param event
   */
  private handleInput = (event: InputEvent & { target: HTMLInputElement }) => {
    this.value = event.target.value;
  };

  /**
   * Handle focus event
   */
  private handleFocus = () => {
    this.classList.add(this.classFocus);
    this.classList = new ClassList(this.classList.classes);
    // Display value
    this.displayValue = this.value;
  };

  /**
   * Handle blur event
   * @param event
   */
  private handleBlur = (event: FocusEvent) => {
    // Manage focus
    this.classList.delete(this.classFocus);
    this.classList = new ClassList(this.classList.classes);
    // Check validity
    this.checkValidity(event.target as HTMLInputElement);
    // Display readonly value
    this.displayValue = this.readonlyValue;
  };

  /**
   * Check if input is valid
   * @param element
   */
  private checkValidity(element:HTMLInputElement) {
    let validity = element.checkValidity();
    // Set error message
    this.errorMessage = undefined;
    // required
    if (!validity && element.validity.valueMissing) {
      this.errorMessage = messages.errors.required;
    }
    // Min & Max
    if (this.min !== undefined && this.numericValue < this.min && this.max === undefined) {
      // Only a min value is set
      validity = false;
      this.errorMessage = messages.errors.currency.min.replace('{min}', `${localeCurrency(this.min)}`);
    } else if (this.max !== undefined && this.numericValue > this.max && this.min === undefined) {
      // Only a max value is set
      validity = false;
      this.errorMessage = messages.errors.currency.max.replace('{max}', `${localeCurrency(this.max)}`);
    } else if ((this.min !== undefined && this.numericValue < this.min) || (this.max !== undefined && this.numericValue > this.max)) {
      // both min and max values are set
      validity = false;
      this.errorMessage = messages.errors.currency.minMax.replace('{min}', `${localeCurrency(this.min)}`).replace('{max}', `${localeCurrency(this.max)}`);
    }

    // Set validity
    this.valid = validity;
    this.invalid = !validity;

    // Update class
    if (validity) {
      this.classList.delete(this.classError);
    } else {
      this.classList.add(this.classError);
    }
  }

  /*************
   * Lifecycle *
   *************/

  componentWillLoad() {
    this.validateValue(this.value);
    // Set displayed value
    this.displayValue = this.readonlyValue;
  }

  /**
   * Check if component props are well configured on init
   */

  render() {
    return (
      <MgInput
        identifier={this.identifier}
        classList={this.classList}
        label={this.label}
        labelOnTop={this.labelOnTop}
        labelColon={this.labelColon}
        required={this.required}
        readonly={this.readonly}
        value={this.value}
        readonlyValue={this.readonlyValue}
        tooltip={this.tooltip}
        displayCharacterLeft={undefined}
        characterLeftTemplate={undefined}
        maxlength={undefined}
        helpText={this.helpText}
        errorMessage={this.errorMessage}
      >
        <input
          type="text"
          value={this.displayValue}
          id={this.identifier}
          name={this.name}
          placeholder={this.placeholder}
          title={this.placeholder}
          disabled={this.disabled}
          required={this.required}
          onInput={this.handleInput}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
      </MgInput>
    );
  }
}
