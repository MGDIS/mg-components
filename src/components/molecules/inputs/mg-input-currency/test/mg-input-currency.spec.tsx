import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInputCurrency } from '../mg-input-currency';
import { messages } from '../../../../../locales';
import { localeCurrency } from '../../../../../utils/locale.utils';

const getPage = (args) => newSpecPage({
  components: [MgInputCurrency],
  template: () => (<mg-input-currency {...args}></mg-input-currency>)
});

describe('mg-input-currency', () => {
  test.each([
    {label: 'label', identifier: "identifier"},
    {label: 'label', identifier: "identifier", labelOnTop: true},
    {label: 'label', identifier: "identifier", readonly: true},
    {label: 'label', identifier: "identifier", readonly: true, value: "1234567890"},
    {label: 'label', identifier: "identifier", tooltip: "My Tooltip Message"},
  ])('Should render with args %s:', async (args) => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each(["", undefined])('Should throw error with invalid label property : %s', async (value) => {
    try {
      await getPage({label:value});
    }
    catch (err) {
      expect(err.message).toMatch('prop "label" is required')
    }
  });

  test('Should trigger events', async ()=> {
    const inputValue = '1234567890';
    const args = {label: 'label', identifier: "identifier", helpText: "My help text"};
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-currency');
    const input = element.shadowRoot.querySelector('input');

    //mock validity
    input.checkValidity = jest.fn(()=> true);
    Object.defineProperty(input, 'validity', { get: jest.fn(()=> ({
      valueMissing: false,
    }))});

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    input.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeTruthy();

    expect(page.root).toMatchSnapshot(); //Snapshot on focus

    input.value = inputValue;
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(parseFloat(inputValue));

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeFalsy();
  });

  test.each([
    {validity: true, valueMissing: false},
    {validity: false, valueMissing: true},
  ])('validity (%s), valueMissing (%s)', async ({validity, valueMissing})=> {
    const args = {label: 'label', identifier: "identifier"};
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-currency');
    const input = element.shadowRoot.querySelector('input');

    //mock validity
    input.checkValidity = jest.fn(()=> validity);
    Object.defineProperty(input, 'validity', { get: jest.fn(()=> ({
      valueMissing,
    }))});

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    if(validity) {
      expect(page.rootInstance.errorMessage).toBeUndefined();
    }
    else if (valueMissing){
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);
    }
    expect(page.rootInstance.valid).toEqual(validity);
    expect(page.rootInstance.invalid).toEqual(!validity);
  });

  test.each([
    {label: "label", identifier: "identifier", min: 10, max: undefined, value: 5},
    {label: "label", identifier: "identifier", min: undefined, max: 10, value: 20},
    {label: "label", identifier: "identifier", min: 10, max: 20, value: 5},
    {label: "label", identifier: "identifier", min: 10, max: 20, value: 25},
  ])('Should return error when value does not match min and max setting (%s)', async (args)=> {
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-currency');
    const input = element.shadowRoot.querySelector('input');

    input.checkValidity = jest.fn(()=> true);

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    if(args.min !== undefined && args.max === undefined) {
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.currency.min.replace('{min}', `${localeCurrency(args.min)}`).replace('{max}', `${localeCurrency(args.max)}`));
    }
    else if(args.min === undefined && args.max !== undefined) {
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.currency.max.replace('{min}', `${localeCurrency(args.min)}`).replace('{max}', `${localeCurrency(args.max)}`));
    }
    else if(args.min !== undefined && args.max !== undefined) {
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.currency.minMax.replace('{min}', `${localeCurrency(args.min)}`).replace('{max}', `${localeCurrency(args.max)}`));
    }

    expect(page.rootInstance.valid).toBeFalsy();
    expect(page.rootInstance.invalid).toBeTruthy();
  });

  test('Should filter entered value', async ()=> {
    let inputValue = '1';
    const args = {label: 'label', identifier: "identifier", helpText: "My help text"};
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-currency');
    const input = element.shadowRoot.querySelector('input');

    //mock validity
    input.checkValidity = jest.fn(()=> true);

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    input.value = 'a';
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(null);

    input.value = inputValue;
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(parseFloat(inputValue));

    input.value = `${inputValue}a`;
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(parseFloat(inputValue));

  });
});
