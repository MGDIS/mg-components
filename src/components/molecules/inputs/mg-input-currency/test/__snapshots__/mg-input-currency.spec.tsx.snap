// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input-currency Should render with args {
  label: 'label',
  identifier: 'identifier',
  readonly: true,
  value: '1234567890'
}: 1`] = `
<mg-input-currency value="1234567890">
  <mock:shadow-root>
    <div class="mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <strong>
          1 234 567 890,00&nbsp;€
        </strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;

exports[`mg-input-currency Should render with args {
  label: 'label',
  identifier: 'identifier',
  tooltip: 'My Tooltip Message'
}: 1`] = `
<mg-input-currency>
  <mock:shadow-root>
    <div class="mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <input aria-describedby="" id="identifier" name="identifier" type="text">
          <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
            <mg-icon icon="info"></mg-icon>
          </mg-tooltip>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;

exports[`mg-input-currency Should render with args { label: 'label', identifier: 'identifier' }: 1`] = `
<mg-input-currency>
  <mock:shadow-root>
    <div class="mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <input aria-describedby="" id="identifier" name="identifier" type="text">
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;

exports[`mg-input-currency Should render with args { label: 'label', identifier: 'identifier', labelOnTop: true }: 1`] = `
<mg-input-currency>
  <mock:shadow-root>
    <div class="is-label-on-top mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <input aria-describedby="" id="identifier" name="identifier" type="text">
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;

exports[`mg-input-currency Should render with args { label: 'label', identifier: 'identifier', readonly: true }: 1`] = `
<mg-input-currency>
  <mock:shadow-root>
    <div class="mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;

exports[`mg-input-currency Should trigger events 1`] = `
<mg-input-currency>
  <mock:shadow-root>
    <div class="is-focused mg-input mg-input--currency">
      <mg-label identifier="identifier">
        label
      </mg-label>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <input aria-describedby="identifier-help-text" id="identifier" name="identifier" type="text">
        </div>
        <mg-help-text identifier="identifier-help-text">
          My help text
        </mg-help-text>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-currency>
`;
